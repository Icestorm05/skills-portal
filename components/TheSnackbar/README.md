# The Snackbar

## Functionality

The snackbar displays a notification message on the bottom of the user's browser screen. By default, the snackbar will stay open for 5 seconds, before emitting a hide event. This component should only be used once within the lifecycle of this application.

## Usage

### Props

`<the-snackbar>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `color` | String *optional* | The color of the snackbar. Defaults to `primary`. Accepts any of: `primary`, `accent`, `secondary`, `info`, `warning`, `error` or `success`.
| `timeout` | Number *optional* | How long the snackbar will stay open for. An event will be emitted once this period elapses.
| `visible` | Boolean | Whether the snackbar is visible or not.

### Events

`<the-snackbar>` emits the following events:

| event | description
| --- | ---
`timeout-elapsed` | Event called when the timeout period has elapsed.

### Slots

`<the-snackbar>` exposes the following slots:

| slot | description
| --- | ---
`default` | The message to display.

### Example

```html
<the-snackbar color="error"
              :timeout="6000"
              :visibility="snackbarVisibility"
              @timeout-elapsed="snackbarVisibility = false">Whoops! Something went wrong.</the-snackbar>
```

```ts
import Component from 'nuxt-class-component';
import Vue from 'vue';

import TheSnackbar from '~/components/TheSnackbar/TheSnackbar.vue';

@Component({
    components: {
        TheSnackbar,
    },
})
export default class GenericComponent extends Vue {
    private snackbarVisibility = true;
}
```