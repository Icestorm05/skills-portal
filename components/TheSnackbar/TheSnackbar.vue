<template lang="pug">
    v-snackbar(:color="color"
               :timeout="timeout"
               :value="visible"
               @input="hide")#TheSnackbar
        span#TheSnackbarMessage
            b(v-if="color === 'error'").mr-1 Error:
            b(v-if="color === 'warning'").mr-1 Warning:
            slot
        v-btn(flat
              aria-label="Close Snackbar"
              @click="hide").ml-3#TheSnackbarCloseButton Close
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import Vue from "vue";
import { Prop, Watch } from "vue-property-decorator";

import VBtn from "vuetify/es5/components/VBtn";
import VIcon from "vuetify/es5/components/VIcon";
import VSnackbar from "vuetify/es5/components/VSnackbar";

@Component({
  components: {
    VBtn,
    VIcon,
    VSnackbar
  }
})
export default class TheSnackbar extends Vue {
  @Prop({ type: String, default: "primary" })
  private color: string;
  @Prop({ type: Number, default: 5000 })
  private timeout: number;
  @Prop({ type: Boolean, default: false })
  private visible: boolean;

  private timer: number = 0;

  @Watch("visible")
  private visibilityChanged(visibility: boolean) {
    if (visibility) {
      this.timer = window.setTimeout(this.hide, this.timeout);
    } else {
      window.clearTimeout(this.timer);
    }
  }

  private hide() {
    this.$emit("timeout-elapsed");
  }
}
</script>
