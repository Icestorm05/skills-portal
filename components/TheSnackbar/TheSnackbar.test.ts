import TheSnackbar from "./TheSnackbar.vue";
import { mount } from "@vue/test-utils";

declare const expect: jest.Expect;

describe("The Snackbar", () => {
  test("is a Vue instance", () => {
    const wrapper = mount(TheSnackbar);
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("snapshot has not changed", () => {
    const wrapper = mount(TheSnackbar);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("error message snapshot has not changed", () => {
    const wrapper = mount(TheSnackbar, {
      propsData: { color: "error", visible: true },
      slots: {
        default: "Test Error Message"
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("success message snapshot has not changed", async () => {
    const wrapper = mount(TheSnackbar, {
      propsData: { color: "success", visible: true },
      slots: {
        default: "Test Success Message"
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("warning message snapshot has not changed", async () => {
    const wrapper = mount(TheSnackbar, {
      propsData: { color: "warning", visible: true },
      slots: {
        default: "Test Success Message"
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("snackbar is hidden by default", () => {
    const wrapper = mount(TheSnackbar);
    expect(wrapper.find("#TheSnackbar").exists()).toBeFalsy();
  });

  test("snackbar is displayed with visible prop", () => {
    const wrapper = mount(TheSnackbar, { propsData: { visible: true } });
    expect(wrapper.find("#TheSnackbar").exists()).toBeTruthy();
  });

  test("color prop changes color of snackbar", () => {
    const wrapper = mount(TheSnackbar, {
      propsData: { color: "warning", visible: true }
    });
    expect(wrapper.find("#TheSnackbar .snack__wrapper").classes()).toContain(
      "warning"
    );
  });

  test("timeout-elapsed event is emitted after timeout prop elapses", async () => {
    const timeout = 1;
    const wrapper = mount(TheSnackbar, {
      propsData: { timeout, visible: true }
    });
    expect(wrapper.emitted("timeout-elapsed")).toBeFalsy();
    await new Promise((resolve, reject) => {
      window.setTimeout(resolve, timeout);
    });
    expect(wrapper.emitted("timeout-elapsed")).toBeTruthy();
  });

  test("visibility watcher is fired when visible prop changes", () => {
    const spy = jest.fn();
    const wrapper = mount(TheSnackbar, {
      propsData: { visible: true },
      watch: {
        visible: spy
      }
    });
    expect(spy).not.toBeCalled();
    wrapper.setProps({ visible: false });
    expect(spy).toBeCalled();
  });

  test("changing the visible prop to true starts a timer", () => {
    const wrapper = mount(TheSnackbar, { propsData: { visible: false } });
    const spy = jest.spyOn(window, "setTimeout");
    expect(spy).not.toBeCalled();
    wrapper.setProps({ visible: true });
    expect(spy).toBeCalled();
  });

  test("changing the visible prop to false cancels a timer", () => {
    const wrapper = mount(TheSnackbar, { propsData: { visible: true } });
    const spy = jest.spyOn(window, "clearTimeout");
    expect(spy).not.toBeCalled();
    wrapper.setProps({ visible: false });
    expect(spy).toBeCalled();
  });

  test("timer property is set with visible prop", () => {
    const wrapper = mount(TheSnackbar, { propsData: { visible: false } });
    expect(wrapper.vm.$data.timer).toBeFalsy();
    wrapper.setProps({ visible: true });
    expect(wrapper.vm.$data.timer).toBeTruthy();
  });

  test("clicking the close snackbar button fires the hide fn", () => {
    const spy = jest.fn();
    const wrapper = mount(TheSnackbar, { propsData: { visible: true } });
    wrapper.setData({ hide: spy });
    expect(spy).not.toBeCalled();
    wrapper.find("#TheSnackbar").trigger("input");
    expect(spy).toBeCalled();
  });

  test("clicking the close snackbar button fires the timeout-elapsed event", () => {
    const wrapper = mount(TheSnackbar, { propsData: { visible: true } });
    expect(wrapper.emitted("timeout-elapsed")).toBeFalsy();
    wrapper.find("#TheSnackbar #TheSnackbarCloseButton").trigger("click");
    expect(wrapper.emitted("timeout-elapsed")).toBeTruthy();
  });
});
