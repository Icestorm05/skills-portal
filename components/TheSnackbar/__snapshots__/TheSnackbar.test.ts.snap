// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Snackbar error message snapshot has not changed 1`] = `
<div class="snack snack--active snack--bottom" id="TheSnackbar">
    <div class="snack__wrapper error">
        <div class="snack__content"><span id="TheSnackbarMessage"><b class="mr-1">Error:</b><!---->Test Error Message</span>
            <button type="button" class="ml-3 btn btn--flat" aria-label="Close Snackbar" id="TheSnackbarCloseButton">
                <div class="btn__content">Close</div>
            </button>
        </div>
    </div>
</div>
`;

exports[`The Snackbar snapshot has not changed 1`] = `undefined`;

exports[`The Snackbar success message snapshot has not changed 1`] = `
<div class="snack snack--active snack--bottom" id="TheSnackbar">
    <div class="snack__wrapper success">
        <div class="snack__content"><span id="TheSnackbarMessage"><!----><!---->Test Success Message</span>
            <button type="button" class="ml-3 btn btn--flat" aria-label="Close Snackbar" id="TheSnackbarCloseButton">
                <div class="btn__content">Close</div>
            </button>
        </div>
    </div>
</div>
`;

exports[`The Snackbar warning message snapshot has not changed 1`] = `
<div class="snack snack--active snack--bottom" id="TheSnackbar">
    <div class="snack__wrapper warning">
        <div class="snack__content"><span id="TheSnackbarMessage"><!----><b class="mr-1">Warning:</b>Test Success Message</span>
            <button type="button" class="ml-3 btn btn--flat" aria-label="Close Snackbar" id="TheSnackbarCloseButton">
                <div class="btn__content">Close</div>
            </button>
        </div>
    </div>
</div>
`;
